// cdktf-cli/bin/cmds/helper/stack-dependencies.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StackDependencies = void 0;
class StackDependencies {
    constructor(stacks) {
        this.stacks = stacks;
        this.pendingStacks = [];
        this.inflightStacks = [];
        this.deployedStacks = [];
        this.pendingStacks = [...this.stacks];
    }
    startRun(stack) {
        this.pendingStacks = this.pendingStacks.filter((item) => item.name !== stack.name);
        this.inflightStacks.push(stack);
    }
    finishRun(stack) {
        this.inflightStacks = this.inflightStacks.filter((item) => item.name !== stack.name);
        this.deployedStacks.push(stack);
    }
    get pendingDeployableStacks() {
        return this.pendingStacks.filter((pendingStack) => {
            const unmetDependencies = pendingStack.dependencies.filter((dependency) => {
                return !this.deployedStacks.some((deployedStack) => {
                    return deployedStack.name === dependency;
                });
            });
            return unmetDependencies.length === 0;
        });
    }
}
exports.StackDependencies = StackDependencies;
