// cdktf-cli/bin/cmds/helper/terraform-check.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.terraformCheck = void 0;
const terraform_cli_1 = require("../../../lib/models/terraform-cli");
const semver = __importStar(require("semver"));
const fs_extra_1 = require("fs-extra");
const path = __importStar(require("path"));
const node_abort_controller_1 = require("node-abort-controller"); // polyfill until we update to node 14
const MIN_SUPPORTED_VERSION = "0.13.0";
const VERSION_REGEXP = /Terraform v\d+.\d+.\d+/;
exports.terraformCheck = async () => {
    try {
        if (fs_extra_1.existsSync(path.join(process.cwd(), "terraform.tfstate"))) {
            throw new Error(`
        CDK for Terraform now supports multiple stacks!
        Found 'terraform.tfstate' Terraform state file. Please rename it to match the stack name. Learn more https://cdk.tf/multiple-stacks
      `);
        }
        // We're abusing the TerraformCli class here,
        // hence we need to construct this object.
        // Only the `workingDirectory` is releveant here.
        const fakeStack = {
            name: "",
            workingDirectory: "./",
            constructPath: "",
            content: "",
            synthesizedStackPath: "",
            annotations: [],
            dependencies: [],
        };
        const terraform = new terraform_cli_1.TerraformCli(new node_abort_controller_1.AbortController().signal, fakeStack);
        const terraformVersion = await terraform.version();
        const terraformVersionMatches = terraformVersion.match(VERSION_REGEXP);
        if (terraformVersionMatches !== null) {
            // Should always be the first match found in the string
            const cleanTerraformVersion = semver.clean(terraformVersionMatches[0].substring(terraformVersionMatches[0].indexOf("v")));
            if (cleanTerraformVersion &&
                semver.lt(cleanTerraformVersion, MIN_SUPPORTED_VERSION)) {
                const warningMessage = `Warning: unsupported Terraform version [${cleanTerraformVersion}] - please upgrade to >=${MIN_SUPPORTED_VERSION}`;
                console.warn(warningMessage);
            }
        }
    }
    catch (e) {
        console.error(e.message);
        process.exit(1);
    }
};
