// @cdktf/provider-generator/lib/get/__tests__/generator/versions-file.test.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const fs = __importStar(require("fs"));
const os = __importStar(require("os"));
const path = __importStar(require("path"));
const constructs_maker_1 = require("../../constructs-maker");
jest.setTimeout(600000);
test("generates a versions file", async () => {
    const workdir = await fs.mkdtempSync(path.join(os.tmpdir(), "versions-file.test"));
    const options = {
        codeMakerOutput: workdir,
        targetLanguage: constructs_maker_1.Language.TYPESCRIPT,
    };
    const constraints = [
        {
            fqn: "hashicorp/aws",
            name: "aws",
            source: "hashicorp/aws",
            version: "4.0.0",
        },
        {
            fqn: "hashicorp/random",
            name: "random",
            source: "hashicorp/random",
            version: "3.1.3",
        },
        {
            fqn: "kreuzwerker/docker",
            name: "docker",
            source: "kreuzwerker/docker",
            version: "2.16.0",
        },
    ];
    const constructMaker = new constructs_maker_1.ConstructsMaker(options, constraints);
    await constructMaker.generate();
    const output = fs.readFileSync(path.join(workdir, "versions.json"), "utf-8");
    expect(Object.keys(JSON.parse(output))).toEqual(expect.arrayContaining(constraints.map((c) => `registry.terraform.io/${c.fqn}`)));
});
