// cdktf-cli/bin/cmds/convert.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const errors_1 = require("../../lib/errors");
const base_command_1 = require("./helper/base-command");
const utilities_1 = require("./helper/utilities");
const path = __importStar(require("path"));
const fs = __importStar(require("fs-extra"));
function readCdktfJson(cwd = process.cwd()) {
    try {
        const cdktfJsonPath = path.join(cwd, "cdktf.json");
        return fs.readJsonSync(cdktfJsonPath);
    }
    catch (e) {
        return undefined;
    }
}
class Command extends base_command_1.BaseCommand {
    constructor() {
        super(...arguments);
        this.command = "convert";
        this.describe = "Converts a single file of HCL configuration to CDK for Terraform. Takes the file to be converted on stdin.";
        this.builder = (args) => {
            var _a;
            return args
                .example("cat main.tf | cdktf convert  --provider integrations/github", "Takes the HCL content of main.tf and converts it to CDK for Terraform content and prints it")
                .example("cat main.tf | cdktf convert --provider hashicorp/aws > imported.ts", "Takes the HCL content of main.tf and converts it to CDK for Terraform content in imported.ts")
                .example("cat main.tf | cdktf convert --provider 'hashicorp/aws@ ~>3.62.0' 'integrations/github@ ~>4.16.0' --language python > imported.py", "Takes the HCL content of main.tf and converts it to CDK for Terraform content in imported.ts")
                .option("language", {
                choices: ["typescript", "python", "csharp", "java"],
                default: ((_a = readCdktfJson()) === null || _a === void 0 ? void 0 : _a.language) || "typescript",
            })
                .option("provider", {
                describe: "The conversion needs to know which providers are used in addition to the ones in your cdktf.json file. We search for a cdktf.json below your current working directory.",
                type: "array",
                default: [],
            })
                .showHelpOnFail(true);
        };
    }
    async handleCommand(argv) {
        errors_1.Errors.setScope("convert");
        // deferred require to keep cdktf-cli main entrypoint small (e.g. for fast shell completions)
        const api = utilities_1.requireHandlers();
        await api.convert(argv);
    }
}
module.exports = new Command();
