// cdktf-cli/bin/cmds/deploy.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const provider_generator_1 = require("@cdktf/provider-generator");
const utilities_1 = require("./helper/utilities");
const errors_1 = require("../../lib/errors");
const base_command_1 = require("./helper/base-command");
const config = provider_generator_1.config.readConfigSync();
class Command extends base_command_1.BaseCommand {
    constructor() {
        super(...arguments);
        this.command = "deploy [stacks...]";
        this.describe = "Deploy the given stacks";
        this.aliases = ["apply"];
        this.builder = (args) => args
            .positional("stacks", {
            desc: "Deploy stacks matching the given ids. Required when more than one stack is present in the app",
            type: "string",
        })
            .option("app", {
            default: config.app,
            required: true,
            desc: "Command to use in order to execute cdktf app",
            alias: "a",
        })
            .option("output", {
            default: config.output,
            required: true,
            desc: "Output directory for the synthesized Terraform config",
            alias: "o",
        })
            .option("auto-approve", {
            type: "boolean",
            default: false,
            required: false,
            desc: "Auto approve",
        })
            .option("outputs-file", {
            type: "string",
            required: false,
            desc: "Path to file where stack outputs will be written as JSON",
            requiresArg: true,
        })
            .option("outputs-file-include-sensitive-outputs", {
            type: "boolean",
            required: false,
            desc: "Whether to include sensitive outputs in the output file",
            default: false,
        })
            .option("ignore-missing-stack-dependencies", {
            type: "boolean",
            required: false,
            desc: "Don't check if all stacks specified in the command have their dependencies included as well",
            default: false,
        })
            .option("parallelism", {
            type: "number",
            required: false,
            desc: "Number of concurrent CDKTF stacks to run. Defaults to infinity, denoted by -1",
            default: -1,
        })
            .option("refresh-only", {
            default: false,
            required: false,
            boolean: true,
            desc: 'Select the "refresh only" planning mode, which checks whether remote objects still match the outcome of the most recent Terraform apply but does not propose any actions to undo any changes made outside of Terraform.',
        })
            .showHelpOnFail(true);
    }
    async handleCommand(argv) {
        errors_1.Errors.setScope("deploy");
        // deferred require to keep cdktf-cli main entrypoint small (e.g. for fast shell completions)
        const api = utilities_1.requireHandlers();
        await api.deploy(argv);
    }
}
module.exports = new Command();
