// cdktf-cli/bin/cmds/get.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const provider_generator_1 = require("@cdktf/provider-generator");
const utilities_1 = require("./helper/utilities");
const errors_1 = require("../../lib/errors");
const base_command_1 = require("./helper/base-command");
const config = provider_generator_1.config.readConfigSync();
class Command extends base_command_1.BaseCommand {
    constructor() {
        super(...arguments);
        this.command = "get";
        this.describe = "Generate CDK Constructs for Terraform providers and modules.";
        this.builder = (args) => args
            .showHelpOnFail(true)
            .option("output", {
            default: config.codeMakerOutput,
            type: "string",
            desc: "Output directory for generated Constructs",
            alias: "o",
        })
            .option("language", {
            default: config.language,
            required: true,
            type: "string",
            desc: "Output programming language",
            alias: "l",
            choices: provider_generator_1.LANGUAGES,
        });
    }
    async handleCommand(argv) {
        errors_1.Errors.setScope("get");
        // deferred require to keep cdktf-cli main entrypoint small (e.g. for fast shell completions)
        const api = utilities_1.requireHandlers();
        await api.get(argv);
    }
}
module.exports = new Command();
