// cdktf-cli/bin/cmds/init.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const init_templates_1 = require("./helper/init-templates");
const utilities_1 = require("./helper/utilities");
const errors_1 = require("../../lib/errors");
const base_command_1 = require("./helper/base-command");
const pkg = utilities_1.readPackageJson();
class Command extends base_command_1.BaseCommand {
    constructor() {
        super(...arguments);
        this.command = "init";
        this.describe = "Create a new cdktf project from a template.";
        this.builder = (args) => args
            .showHelpOnFail(true)
            .option("template", {
            type: "string",
            desc: `The template to be used to create a new project. Either URL to zip file or one of the built-in templates: [${init_templates_1.templates
                .map((t) => `"${t}"`)
                .join(", ")}]`,
        })
            .option("project-name", {
            type: "string",
            desc: "The name of the project.",
        })
            .option("project-description", {
            type: "string",
            desc: "The description of the project.",
        })
            .option("dist", {
            type: "string",
            desc: 'Install dependencies from a "dist" directory (for development)',
        })
            .option("local", {
            type: "boolean",
            desc: "Use local state storage for generated Terraform.",
            default: false,
        })
            .option("cdktf-version", {
            type: "string",
            desc: "The cdktf version to use while creating a new project.",
            default: pkg.version,
        })
            .option("from-terraform-project", {
            type: "string",
            desc: "Use a terraform project as the basis, CDK constructs will be generated based on the .tf files in the path",
        })
            .option("enable-crash-reporting", {
            type: "boolean",
            desc: "Enable crash reporting for the CLI, see https://www.terraform.io/cdktf/telemetry#crash-reporting for more details",
        })
            .strict();
    }
    async handleCommand(argv) {
        errors_1.Errors.setScope("init");
        // deferred require to keep cdktf-cli main entrypoint small (e.g. for fast shell completions)
        const api = utilities_1.requireHandlers();
        await api.init(argv);
    }
}
module.exports = new Command();
