// cdktf-cli/bin/cmds/login.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const utilities_1 = require("./helper/utilities");
const errors_1 = require("../../lib/errors");
const base_command_1 = require("./helper/base-command");
class Command extends base_command_1.BaseCommand {
    constructor() {
        super(...arguments);
        this.command = "login";
        this.describe = "Retrieves an API token to connect to Terraform Cloud or Terraform Enterprise.";
        this.builder = (args) => args
            .showHelpOnFail(true)
            .option("tfe-hostname", {
            string: true,
            requiresArg: true,
            describe: "The Terraform Enterprise hostname to authenticate against. If you use Terraform Cloud you can leave this on the default.",
            default: "app.terraform.io",
        })
            .example("cdktf login", "Takes you through the interactive login process")
            .example("cdktf login --tfe-hostname tfe.my-company.com", "Takes you through the interactive login process on your companies Terraform Enterprise instance.")
            .example("cat my-token.txt | cdktf login", "Uses a locally stored token directly, instead of going through the interactive login process")
            .strict();
    }
    async handleCommand(argv) {
        errors_1.Errors.setScope("login");
        // deferred require to keep cdktf-cli main entrypoint small (e.g. for fast shell completions)
        const api = utilities_1.requireHandlers();
        await api.login(argv);
    }
}
module.exports = new Command();
