// cdktf-cli/bin/cmds/output.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const provider_generator_1 = require("@cdktf/provider-generator");
const utilities_1 = require("./helper/utilities");
const errors_1 = require("../../lib/errors");
const base_command_1 = require("./helper/base-command");
const config = provider_generator_1.config.readConfigSync();
class Command extends base_command_1.BaseCommand {
    constructor() {
        super(...arguments);
        this.command = "output [stacks..]";
        this.describe = "Prints the output of stacks";
        this.aliases = ["outputs"];
        this.builder = (args) => args
            .positional("stacks", {
            desc: "Get outputs of the stacks matching the given ids. Required when more than one stack is present in the app",
            type: "string",
        })
            .option("app", {
            default: config.app,
            required: true,
            desc: "Command to use in order to execute cdktf app",
            alias: "a",
        })
            .option("output", {
            default: config.output,
            required: true,
            desc: "Output directory for the synthesized Terraform config",
            alias: "o",
        })
            .option("outputs-file", {
            type: "string",
            required: false,
            desc: "Path to file where stack outputs will be written as JSON",
            requiresArg: true,
        })
            .option("outputs-file-include-sensitive-outputs", {
            type: "boolean",
            required: false,
            desc: "Whether to include sensitive outputs in the output file",
            default: false,
        })
            .showHelpOnFail(true);
    }
    async handleCommand(argv) {
        errors_1.Errors.setScope("output");
        // deferred require to keep cdktf-cli main entrypoint small (e.g. for fast shell completions)
        const api = utilities_1.requireHandlers();
        await api.output(argv);
    }
}
module.exports = new Command();
