// cdktf-cli/bin/cmds/provider-add.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const utilities_1 = require("./helper/utilities");
const errors_1 = require("../../lib/errors");
const base_command_1 = require("./helper/base-command");
class Command extends base_command_1.BaseCommand {
    constructor() {
        super(...arguments);
        this.command = "add <provider...>";
        this.describe = "Add one or more Terraform providers to your project.";
        this.builder = (args) => args
            .showHelpOnFail(true)
            .positional("provider", {
            type: "string",
            desc: "Name of the provider to add. Can include a version constraint (e.g. aws@~>4.0).",
            required: true,
        })
            .option("force-local", {
            type: "boolean",
            desc: "force local provider installation, even if pre-built provider exists",
        });
    }
    async handleCommand(argv) {
        errors_1.Errors.setScope("provider add");
        // deferred require to keep cdktf-cli main entrypoint small (e.g. for fast shell completions)
        const api = utilities_1.requireHandlers();
        await api.providerAdd(argv);
    }
}
module.exports = new Command();
