// cdktf-cli/bin/cmds/synth.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const provider_generator_1 = require("@cdktf/provider-generator");
const utilities_1 = require("./helper/utilities");
const errors_1 = require("../../lib/errors");
const base_command_1 = require("./helper/base-command");
const config = provider_generator_1.config.readConfigSync();
class Command extends base_command_1.BaseCommand {
    constructor() {
        super(...arguments);
        this.command = "synth";
        this.describe = "Synthesizes Terraform code for the given app in a directory.";
        this.aliases = ["synthesize"];
        this.builder = (args) => args
            .option("app", {
            default: config.app,
            desc: "Command to use in order to execute cdktf app",
            alias: "a",
        })
            .option("output", {
            default: config.output,
            desc: "Output directory for the synthesized Terraform config",
            alias: "o",
        })
            .option("check-code-maker-output", {
            type: "boolean",
            desc: "Should `codeMakerOutput` existence check be performed? By default it will be checked if providers or modules are configured.",
            default: provider_generator_1.config.shouldCheckCodeMakerOutput(config),
        })
            .showHelpOnFail(true);
    }
    async handleCommand(argv) {
        errors_1.Errors.setScope("synth");
        // deferred require to keep cdktf-cli main entrypoint small (e.g. for fast shell completions)
        const api = utilities_1.requireHandlers();
        await api.synth(argv);
    }
}
module.exports = new Command();
