// cdktf/lib/validations/validate-provider-presence.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ValidateProviderPresence = void 0;
const constructs_1 = require("constructs");
const terraform_provider_1 = require("../terraform-provider");
const terraform_resource_1 = require("../terraform-resource");
const terraform_data_source_1 = require("../terraform-data-source");
class ValidateProviderPresence {
    constructor(host) {
        this.host = host;
        this.providerNames = new Set();
        this.foundProviders = [];
    }
    check(node) {
        if (node instanceof terraform_resource_1.TerraformResource ||
            node instanceof terraform_data_source_1.TerraformDataSource) {
            if (node.terraformGeneratorMetadata) {
                this.providerNames.add(node.terraformGeneratorMetadata.providerName);
            }
        }
        if (node instanceof terraform_provider_1.TerraformProvider) {
            this.foundProviders.push(node);
        }
        for (const child of constructs_1.Node.of(node).children) {
            this.check(child);
        }
    }
    validate() {
        this.check(this.host);
        const missingProviders = Array.from(this.providerNames).filter((name) => {
            return !this.foundProviders.some((p) => p.terraformGeneratorMetadata?.providerName === name);
        });
        if (missingProviders.length === 0) {
            return [];
        }
        else {
            return [
                `Found resources without a matching provider. Please make sure to add the following providers to your stack: ${missingProviders.join(", ")}`,
            ];
        }
    }
}
exports.ValidateProviderPresence = ValidateProviderPresence;
