// cdktf-cli/lib/cdktf-config.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CdktfConfig = void 0;
const fs = __importStar(require("fs"));
const provider_generator_1 = require("@cdktf/provider-generator");
const utilities_1 = require("../bin/cmds/helper/utilities");
const errors_1 = require("./errors");
const path_1 = __importDefault(require("path"));
const logging_1 = require("./logging");
const config_1 = require("@cdktf/provider-generator/lib/config");
// TODO: move this to some common package from where it is used by @cdktf/provider-generator, cdktf, and cdktf-cli
// (might end up in cdktf)
// tracked here https://github.com/hashicorp/terraform-cdk/issues/1814
class CdktfConfig {
    constructor(cdktfConfigPath) {
        this.cdktfConfigPath = cdktfConfigPath;
    }
    readCdktfConfig() {
        const cdktfConfig = require(this.cdktfConfigPath);
        if (typeof cdktfConfig !== "object" || cdktfConfig === null) {
            throw errors_1.Errors.External("cdktf.json is malformed. The root must be a JSON object."); // TODO: define a schema and validate against it
        }
        return cdktfConfig;
    }
    writeCdktfConfig(cdktfConfig) {
        const cdktfConfigString = JSON.stringify(cdktfConfig, null, 2);
        fs.writeFileSync(this.cdktfConfigPath, cdktfConfigString);
    }
    getProperty(property) {
        const cdktfConfig = this.readCdktfConfig();
        return cdktfConfig[property] || config_1.CONFIG_DEFAULTS[property];
    }
    get language() {
        const rawLanguage = this.getProperty("language");
        const language = Object.values(provider_generator_1.Language).find((value) => value === rawLanguage);
        if (!language)
            throw errors_1.Errors.Usage(`${rawLanguage} is not a valid language. It must be one of ${Object.values(provider_generator_1.Language)}`);
        return language;
    }
    get codeMakerOutput() {
        return this.getProperty("codeMakerOutput");
    }
    get terraformProviders() {
        const providers = this.getProperty("terraformProviders");
        if (!Array.isArray(providers))
            return [];
        return providers;
    }
    writeTerraformProviders(providers) {
        const cdktfConfig = this.readCdktfConfig();
        cdktfConfig.terraformProviders = providers;
        this.writeCdktfConfig(cdktfConfig);
    }
    get projectDirectory() {
        return path_1.default.dirname(this.cdktfConfigPath);
    }
    static read(path = process.cwd()) {
        const cdktfConfigPath = utilities_1.findFileAboveCwd("cdktf.json", path);
        if (!cdktfConfigPath) {
            throw errors_1.Errors.External("Could not find cdktf.json. Make sure there is a cdktf.json file in the current directory or one of its parents.");
        }
        logging_1.logger.trace(`cdktf.json found at ${cdktfConfigPath}`);
        return new CdktfConfig(cdktfConfigPath);
    }
}
exports.CdktfConfig = CdktfConfig;
