// cdktf-cli/lib/errors.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Errors = exports.IsErrorType = void 0;
const checkpoint_1 = require("./checkpoint");
const version_1 = require("./version");
const Sentry = __importStar(require("@sentry/node"));
// Errors that will emit telemetry events
async function report(command, payload) {
    const reportParams = {
        command,
        product: "cdktf",
        version: `${version_1.DISPLAY_VERSION}`,
        dateTime: new Date(),
        payload,
    };
    await checkpoint_1.ReportRequest(reportParams);
}
function IsErrorType(error, type) {
    return error && error.__type === type;
}
exports.IsErrorType = IsErrorType;
function reportPrefixedError(type, command) {
    return (message, context) => {
        report(command, { ...context, message, type });
        const err = new Error(`${type} Error: ${message}`);
        Object.entries(context || {}).forEach(([key, value]) => {
            err[key] = value;
        });
        err.__type = type;
        return err;
    };
}
// The CLI only deals with one command at a time, so we can just use the same
// scope for all errors and set it once during initialization.
let errorScope = "unknown";
exports.Errors = {
    // Error within our control
    Internal: reportPrefixedError("Internal", errorScope),
    // Error in the usage
    Usage: reportPrefixedError("Usage", errorScope),
    // Error outside of our control (e.g. terraform failed)
    External: reportPrefixedError("External", errorScope),
    // Set the scope for all errors
    setScope(scope) {
        errorScope = scope;
        Sentry.configureScope((s) => s.setTransactionName(scope));
    },
};
