// cdktf-cli/lib/logging.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.logFileName = exports.processLoggerError = exports.processLoggerDebug = exports.logger = void 0;
const log4js_1 = require("log4js");
const fs = __importStar(require("fs-extra"));
const path = __importStar(require("path"));
const Sentry = __importStar(require("@sentry/node"));
const provider_generator_1 = require("@cdktf/provider-generator");
const hcl2cdk_1 = require("@cdktf/hcl2cdk");
const cliLogger = log4js_1.getLogger();
const logger = {
    trace(message, ...args) {
        cliLogger.trace(message, ...args);
        Sentry.addBreadcrumb({
            message,
            level: Sentry.Severity.Debug,
        });
    },
    debug(message, ...args) {
        cliLogger.debug(message, ...args);
        Sentry.addBreadcrumb({
            message,
            level: Sentry.Severity.Debug,
        });
    },
    info(message, ...args) {
        cliLogger.info(message, ...args);
        Sentry.addBreadcrumb({
            message,
            level: Sentry.Severity.Info,
        });
    },
    warn(message, ...args) {
        cliLogger.warn(message, ...args);
        Sentry.addBreadcrumb({
            message,
            level: Sentry.Severity.Warning,
        });
    },
    error(message, ...args) {
        cliLogger.error(message, ...args);
        Sentry.addBreadcrumb({
            message,
            level: Sentry.Severity.Error,
        });
    },
    fatal(message, ...args) {
        cliLogger.fatal(message, ...args);
        Sentry.addBreadcrumb({
            message,
            level: Sentry.Severity.Critical,
        });
    },
};
exports.logger = logger;
cliLogger.level = process.env.CDKTF_LOG_LEVEL || "INFO";
const logFileName = "cdktf.log";
exports.logFileName = logFileName;
if (process.env.CDKTF_LOG_FILE_DIRECTORY &&
    fs.existsSync(process.env.CDKTF_LOG_FILE_DIRECTORY) &&
    fs.statSync(process.env.CDKTF_LOG_FILE_DIRECTORY).isDirectory()) {
    log4js_1.configure({
        appenders: {
            cdktf: {
                type: "file",
                filename: path.join(process.env.CDKTF_LOG_FILE_DIRECTORY, logFileName),
            },
        },
        categories: { default: { appenders: ["cdktf"], level: "debug" } },
    });
}
const processLoggerDebug = (chunk) => {
    logger.debug(chunk.toString());
};
exports.processLoggerDebug = processLoggerDebug;
const processLoggerError = (chunk) => {
    logger.error(chunk.toString());
};
exports.processLoggerError = processLoggerError;
// We have this mechanism to allow loggers from sub-packages to hook into this logging mechanism
hcl2cdk_1.setLogger(logger);
provider_generator_1.setLogger(logger);
