// @cdktf/provider-generator/lib/__tests__/provider.test.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const path = __importStar(require("path"));
const fs = __importStar(require("fs-extra"));
const glob_1 = require("glob");
const util_1 = require("../util");
const constructs_maker_1 = require("../get/constructs-maker");
const config_1 = require("../config");
function directorySnapshot(root) {
    const output = {};
    const files = glob_1.glob.sync("**", {
        ignore: [".git/**", ".jsii"],
        cwd: root,
        nodir: true,
        dot: true,
    });
    for (const file of files) {
        const filePath = path.join(root, file);
        let content;
        if (path.extname(filePath) === ".json") {
            content = fs.readJsonSync(filePath);
        }
        else {
            content = fs.readFileSync(filePath, "utf-8");
        }
        output[file] = content;
    }
    return output;
}
describe("Provider", () => {
    it("generates a provider", async () => {
        const constraint = new config_1.TerraformProviderConstraint("DataDog/datadog@= 3.12.0");
        return await util_1.mkdtemp(async (workdir) => {
            const jsiiPath = path.join(workdir, ".jsii");
            const maker = new constructs_maker_1.ConstructsMaker({
                codeMakerOutput: workdir,
                outputJsii: jsiiPath,
                targetLanguage: constructs_maker_1.Language.TYPESCRIPT,
            }, [constraint]);
            await maker.generate();
            const snapshot = directorySnapshot(workdir);
            expect(snapshot).toMatchSnapshot();
        });
    }, 600000);
});
