// @cdktf/provider-generator/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !exports.hasOwnProperty(p)) __createBinding(exports, m, p);
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateProviderBindingsFromSchema = exports.CodeMaker = exports.TerraformProviderGenerator = void 0;
__exportStar(require("./get/constructs-maker"), exports);
exports.config = __importStar(require("./config"));
var provider_namespaces_1 = require("./get/generator/constants/provider-namespaces");
Object.defineProperty(exports, "getResourceNamespace", { enumerable: true, get: function () { return provider_namespaces_1.getResourceNamespace; } });
var module_1 = require("./get/module");
Object.defineProperty(exports, "isRegistryModule", { enumerable: true, get: function () { return module_1.isRegistryModule; } });
var config_1 = require("./config");
Object.defineProperty(exports, "TerraformModuleConstraint", { enumerable: true, get: function () { return config_1.TerraformModuleConstraint; } });
Object.defineProperty(exports, "isLocalModule", { enumerable: true, get: function () { return config_1.isLocalModule; } });
var provider_schema_1 = require("./get/generator/provider-schema");
Object.defineProperty(exports, "readSchema", { enumerable: true, get: function () { return provider_schema_1.readSchema; } });
const codemaker_1 = require("codemaker");
Object.defineProperty(exports, "CodeMaker", { enumerable: true, get: function () { return codemaker_1.CodeMaker; } });
const constructs_maker_1 = require("./get/constructs-maker");
var models_1 = require("./get/generator/models");
Object.defineProperty(exports, "escapeAttributeName", { enumerable: true, get: function () { return models_1.escapeAttributeName; } });
const provider_generator_1 = require("./get/generator/provider-generator");
Object.defineProperty(exports, "TerraformProviderGenerator", { enumerable: true, get: function () { return provider_generator_1.TerraformProviderGenerator; } });
var config_2 = require("./config");
Object.defineProperty(exports, "setLogger", { enumerable: true, get: function () { return config_2.setLogger; } });
// Used for testing only
async function generateProviderBindingsFromSchema(targetPath, schemaJSON, options) {
    const code = new codemaker_1.CodeMaker();
    const generator = new provider_generator_1.TerraformProviderGenerator(code, schemaJSON);
    generator.generateAll();
    await code.save(targetPath);
    if (options) {
        await constructs_maker_1.generateJsiiLanguage(code, options);
    }
}
exports.generateProviderBindingsFromSchema = generateProviderBindingsFromSchema;
