// cdktf-cli/bin/cmds/ui/deploy.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Deploy = exports.DeploySummary = void 0;
/* eslint-disable no-control-regex */
const react_1 = __importStar(require("react"));
const ink_1 = require("ink");
const cdktf_project_1 = require("./hooks/cdktf-project");
const components_1 = require("./components");
exports.DeploySummary = ({ resources, }) => {
    const summary = resources.reduce((accumulator, resource) => {
        if (accumulator[resource.applyState] !== undefined) {
            accumulator[resource.applyState] += 1;
        }
        return accumulator;
    }, {
        created: 0,
        updated: 0,
        destroyed: 0,
    });
    return (react_1.default.createElement(react_1.default.Fragment, null, Object.keys(summary).map((key, i) => (react_1.default.createElement(ink_1.Box, { key: key },
        i > 0 && react_1.default.createElement(ink_1.Text, null, ", "),
        react_1.default.createElement(ink_1.Text, null,
            summary[key],
            " ",
            key))))));
};
exports.Deploy = ({ outDir, targetStacks, synthCommand, autoApprove, onOutputsRetrieved, outputsPath, ignoreMissingStackDependencies, parallelism, refreshOnly, }) => {
    const [outputs, setOutputs] = react_1.useState();
    const { status, logEntries } = cdktf_project_1.useCdktfProject({ outDir, synthCommand }, async (project) => {
        await project.deploy({
            stackNames: targetStacks,
            autoApprove,
            ignoreMissingStackDependencies,
            parallelism,
            refreshOnly,
        });
        if (onOutputsRetrieved) {
            onOutputsRetrieved(project.outputsByConstructId);
        }
        setOutputs(project.outputsByConstructId);
    });
    const bottomBar = status.type === "done" ? (react_1.default.createElement(components_1.OutputsBottomBar, { outputs: outputs, outputsFile: outputsPath })) : (status === null || status === void 0 ? void 0 : status.type) === "waiting for approval of stack" ? (react_1.default.createElement(components_1.ApproveBottomBar, { stackName: status.stackName, onApprove: status.approve, onDismiss: status.dismiss, onStop: status.stop })) : (react_1.default.createElement(components_1.ExecutionStatusBottomBar, { status: status, actionName: "deploying" }));
    return react_1.default.createElement(components_1.StreamView, { logs: logEntries }, bottomBar);
};
