// cdktf-cli/bin/cmds/ui/get.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Get = void 0;
const react_1 = __importStar(require("react"));
const ink_1 = require("ink");
const ink_spinner_1 = __importDefault(require("ink-spinner"));
const provider_generator_1 = require("@cdktf/provider-generator");
const checkpoint_1 = require("../../../lib/checkpoint");
const lib_1 = require("../../../lib");
exports.Get = ({ codeMakerOutput, language, constraints, }) => {
    const [currentStatus, setCurrentStatus] = react_1.default.useState(lib_1.GetStatus.STARTING);
    const { exit } = ink_1.useApp();
    const constructsOptions = {
        codeMakerOutput: codeMakerOutput,
        targetLanguage: language,
    };
    react_1.default.useEffect(() => {
        const runGet = async () => {
            try {
                await lib_1.get({
                    constraints,
                    constructsOptions,
                    onUpdate: setCurrentStatus,
                    reportTelemetry: (payload) => checkpoint_1.sendTelemetry("get", {
                        language: payload.targetLanguage,
                        ...payload.trackingPayload,
                    }),
                });
            }
            catch (e) {
                console.error(e);
                exit(e);
            }
        };
        runGet();
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, []); // only once, we don't expect props to change
    react_1.default.useEffect(() => {
        if (currentStatus === lib_1.GetStatus.DONE) {
            exit();
        }
        if (currentStatus === lib_1.GetStatus.ERROR) {
            exit(new Error(`ERROR: synthesis failed, app expected to create "${codeMakerOutput}"`));
        }
    }, [currentStatus]);
    const isGenerating = currentStatus != lib_1.GetStatus.DONE;
    const statusText = `${currentStatus}...`;
    const jsonTerraformOutput = (react_1.default.createElement(ink_1.Text, null,
        "Generated ",
        react_1.default.createElement(ink_1.Text, { color: "green" }, language),
        " constructs in the output directory: ",
        react_1.default.createElement(ink_1.Text, { bold: true }, codeMakerOutput),
        language === provider_generator_1.Language.GO && (react_1.default.createElement(react_1.default.Fragment, null,
            react_1.default.createElement(ink_1.Newline, null),
            react_1.default.createElement(ink_1.Newline, null),
            react_1.default.createElement(ink_1.Text, null,
                "The generated code depends on",
                " ",
                react_1.default.createElement(ink_1.Text, { color: "cyan" }, "jsii-runtime-go"),
                ". If you haven't yet installed it, you can run",
                " ",
                react_1.default.createElement(ink_1.Text, { color: "blueBright" }, "go mod tidy"),
                " to automatically install it.")))));
    return (react_1.default.createElement(ink_1.Box, null, isGenerating ? (react_1.default.createElement(react_1.Fragment, null,
        react_1.default.createElement(ink_1.Text, { color: "green" },
            react_1.default.createElement(ink_spinner_1.default, { type: "dots" })),
        react_1.default.createElement(ink_1.Box, { paddingLeft: 1 },
            react_1.default.createElement(ink_1.Text, null, statusText)))) : (react_1.default.createElement(react_1.Fragment, null,
        react_1.default.createElement(ink_1.Box, null, jsonTerraformOutput)))));
};
