// cdktf-cli/bin/cmds/ui/watch.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Watch = void 0;
const node_abort_controller_1 = require("node-abort-controller");
const react_1 = __importStar(require("react"));
const lib_1 = require("../../../lib/");
const components_1 = require("./components");
exports.Watch = ({ targetDir, targetStacks, synthCommand, autoApprove, parallelism, }) => {
    const [logEntryId, setLogEntryId] = react_1.useState(0);
    const [logEntries, setLogEntries] = react_1.useState([]);
    const [currentState, setCurrentState] = react_1.useState({
        type: "waiting",
    });
    react_1.useEffect(() => {
        const ac = new node_abort_controller_1.AbortController();
        const onAbort = () => {
            ac.abort();
        };
        process.on("SIGINT", onAbort);
        process.on("SIGTERM", onAbort);
        process.on("SIGQUIT", onAbort);
        lib_1.watch({
            synthCommand,
            outDir: targetDir,
            onUpdate: () => { },
            onLog: ({ stackName, message, messageWithConstructPath }) => {
                setLogEntries((prev) => [
                    ...prev,
                    {
                        id: `${stackName}-${logEntryId}`,
                        stackName: stackName,
                        content: messageWithConstructPath
                            ? messageWithConstructPath
                            : message,
                    },
                ]);
                setLogEntryId((current) => current + 1);
            },
        }, {
            autoApprove,
            stackNames: targetStacks,
            parallelism,
        }, ac.signal, (state) => {
            setCurrentState(state);
        });
    }, []);
    return (react_1.default.createElement(components_1.StreamView, { logs: logEntries },
        react_1.default.createElement(components_1.WatchStatusBottomBar, { currentState: currentState })));
};
