// @cdktf/provider-generator/lib/get/generator/provider-generator.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TerraformProviderGenerator = void 0;
const resource_parser_1 = require("./resource-parser");
const emitter_1 = require("./emitter");
const constructs_maker_1 = require("../constructs-maker");
const path = __importStar(require("path"));
const config_1 = require("../../config");
const isMatching = (target, terraformSchemaName) => {
    if (target.isModule)
        return false;
    const elements = terraformSchemaName.split("/");
    if (elements.length === 1) {
        return target.source === terraformSchemaName;
    }
    else {
        const [hostname, scope, provider] = elements;
        if (!hostname || !scope || !provider) {
            throw new Error(`can't handle ${terraformSchemaName}`);
        }
        return target.name === provider;
    }
};
class TerraformProviderGenerator {
    constructor(code, schema) {
        this.code = code;
        this.schema = schema;
        this.resourceParser = new resource_parser_1.ResourceParser();
        this.versions = {};
        this.code.indentation = 2;
        this.resourceEmitter = new emitter_1.ResourceEmitter(this.code);
        this.structEmitter = new emitter_1.StructEmitter(this.code);
    }
    getProviderByConstraint(providerConstraint) {
        return Object.keys(this.schema.provider_schemas || {}).find((fqpn) => isMatching(providerConstraint, fqpn));
    }
    generate(providerConstraint) {
        var _a;
        const fqpn = this.getProviderByConstraint(providerConstraint);
        if (!fqpn) {
            config_1.logger.debug(`Could not find provider constraint for ${providerConstraint} in schema: ${JSON.stringify(this.schema, null, 2)}`);
            throw new Error(`Could not find provider with constraint ${providerConstraint}`);
        }
        const providerVersion = (_a = this.schema.provider_versions) === null || _a === void 0 ? void 0 : _a[fqpn];
        this.emitProvider(fqpn, providerVersion, providerConstraint);
        this.versions[fqpn] = providerVersion;
    }
    generateAll() {
        for (const fqpn of Object.keys(this.schema.provider_schemas || {})) {
            this.generate(new constructs_maker_1.ConstructsMakerProviderTarget(new config_1.TerraformProviderConstraint(fqpn), constructs_maker_1.LANGUAGES[0]));
        }
    }
    async save(outdir) {
        await this.code.save(outdir);
    }
    buildResourceModels(fqpn) {
        var _a;
        const name = fqpn.split("/").pop();
        if (!name) {
            throw new Error(`can't handle ${fqpn}`);
        }
        const provider = (_a = this.schema.provider_schemas) === null || _a === void 0 ? void 0 : _a[fqpn];
        if (!provider) {
            throw new Error(`Can not find provider '${fqpn}' in schema`);
        }
        const resources = Object.entries(provider.resource_schemas || {}).map(([type, resource]) => this.resourceParser.parse(name, type, resource, "resource"));
        const dataSources = Object.entries(provider.data_source_schemas || {}).map(([type, resource]) => this.resourceParser.parse(name, `data_${type}`, resource, "data_source"));
        return [].concat(...resources, ...dataSources);
    }
    getClassNameForResource(terraformType) {
        return this.resourceParser.getClassNameForResource(terraformType);
    }
    emitProvider(fqpn, providerVersion, constraint) {
        var _a;
        const name = fqpn.split("/").pop();
        if (!name) {
            throw new Error(`can't handle ${fqpn}`);
        }
        const provider = (_a = this.schema.provider_schemas) === null || _a === void 0 ? void 0 : _a[fqpn];
        if (!provider) {
            throw new Error(`Can not find provider '${fqpn}' in schema`);
        }
        const namespacedResources = {};
        const files = [];
        this.buildResourceModels(fqpn).forEach((resourceModel) => {
            if (constraint) {
                resourceModel.providerVersionConstraint = constraint.version;
                resourceModel.terraformProviderSource = constraint.source;
            }
            resourceModel.providerVersion = providerVersion;
            if (resourceModel.namespace) {
                const namespace = resourceModel.namespace.name;
                if (!namespacedResources[namespace]) {
                    namespacedResources[namespace] = [];
                }
                namespacedResources[namespace].push(resourceModel);
            }
            else if (resourceModel.structsRequireSharding) {
                files.push(this.emitResourceFileWithComplexStruct(resourceModel));
            }
            else {
                files.push(this.emitResourceFile(resourceModel));
            }
        });
        for (const [, resources] of Object.entries(namespacedResources)) {
            files.push(this.emitNamespacedResourceFile(name, resources));
        }
        if (provider.provider) {
            const providerResource = this.resourceParser.parse(name, `provider`, provider.provider, "provider");
            if (constraint) {
                providerResource.providerVersionConstraint = constraint.version;
                providerResource.terraformProviderSource = constraint.source;
            }
            providerResource.providerVersion = providerVersion;
            files.push(this.emitResourceFile(providerResource));
        }
        this.emitIndexFile(name, files);
    }
    emitIndexFile(provider, files) {
        const folder = `providers/${provider}`;
        const filePath = `${folder}/index.ts`;
        this.code.openFile(filePath);
        this.code.line("// generated by cdktf get");
        for (const file of files) {
            if (file.startsWith("ns:")) {
                const fileName = file.replace("ns:", "");
                this.code.line(`export * as ${fileName} from './${fileName}'`);
            }
            else {
                this.code.line(`export * from './${file
                    .replace(`${folder}/`, "")
                    .replace(".ts", "")}';`);
            }
        }
        this.code.line();
        this.code.closeFile(filePath);
    }
    emitResourceFile(resource) {
        this.code.openFile(resource.filePath);
        this.emitFileHeader(resource);
        this.structEmitter.emit(resource);
        this.resourceEmitter.emit(resource);
        this.code.closeFile(resource.filePath);
        return resource.filePath;
    }
    emitNamespacedResourceFile(providerName, resources) {
        const ns = resources[0].namespace;
        const comment = ns === null || ns === void 0 ? void 0 : ns.comment;
        if (!(ns === null || ns === void 0 ? void 0 : ns.name))
            throw new Error("namespace name is missing");
        const generatedFiles = [];
        for (const resource of resources) {
            // drop the last segment of the filepath
            const filePath = resource.filePath.split("/").slice(0, -1).join("/");
            const namespacedFilePath = path.join(filePath, ns.name, resource.fileName);
            this.code.openFile(namespacedFilePath);
            this.code.line(`// generated from terraform resource schema`);
            this.code.line();
            if (resource.structsRequireSharding) {
                this.code.line(`import { ${resource.referencedTypes.join(", \n")}} from './${resource.structsFolderName}'`);
                resource.importStatements.forEach((statement) => this.code.line(statement));
                this.code.line();
                this.code.line(`/**`);
                this.code.line(`* ${comment}`);
                this.code.line(`*/`);
                this.structEmitter.emitInterface(resource, resource.configStruct);
                this.resourceEmitter.emit(resource);
                this.code.closeFile(namespacedFilePath);
                this.structEmitter.emit(resource);
                generatedFiles.push(resource.fileName);
                generatedFiles.push(resource.structsFolderName);
            }
            else {
                resource.importStatements.forEach((statement) => this.code.line(statement));
                this.code.line();
                this.code.line(`/**`);
                this.code.line(`* ${comment}`);
                this.code.line(`*/`);
                this.structEmitter.emit(resource);
                this.resourceEmitter.emit(resource);
                this.code.closeFile(namespacedFilePath);
                generatedFiles.push(resource.fileName);
            }
        }
        const indexFilePath = path.join(`providers`, providerName, ns.name, "index.ts");
        this.code.openFile(indexFilePath);
        this.code.line("// generated by cdktf get");
        for (const file of generatedFiles) {
            this.code.line(`export * from './${path.basename(file, ".ts")}';`);
        }
        this.code.line();
        this.code.closeFile(indexFilePath);
        return `ns:${ns.name}`;
    }
    emitResourceFileWithComplexStruct(resource) {
        const generatedFiles = [];
        // drop the last segment of the filepath
        const filePath = resource.filePath;
        this.code.openFile(filePath);
        this.code.line(`// generated from terraform resource schema`);
        this.code.line();
        if (resource.structsRequireSharding) {
            if (resource.referencedTypes.length > 0) {
                this.code.line(`import { ${resource.referencedTypes.join(", \n")}} from './${resource.structsFolderName}'`);
            }
            this.code.line(`export * from './${resource.structsFolderName}'`);
            resource.importStatements.forEach((statement) => this.code.line(statement));
            this.structEmitter.emitInterface(resource, resource.configStruct);
            this.resourceEmitter.emit(resource);
            this.code.closeFile(filePath);
            this.structEmitter.emit(resource);
            generatedFiles.push(resource.fileName);
            generatedFiles.push(resource.structsFolderName);
        }
        else {
            resource.importStatements.forEach((statement) => this.code.line(statement));
            this.structEmitter.emit(resource);
            this.resourceEmitter.emit(resource);
            this.code.closeFile(filePath);
            generatedFiles.push(resource.fileName);
        }
        return filePath;
    }
    emitFileHeader(resource) {
        this.code.line(`// ${resource.linkToDocs}`);
        this.code.line(`// generated from terraform resource schema`);
        this.code.line();
        resource.importStatements.forEach((statement) => this.code.line(statement));
        this.code.line();
        this.code.line("// Configuration");
        this.code.line();
    }
}
exports.TerraformProviderGenerator = TerraformProviderGenerator;
