// cdktf-cli/bin/cmds/ui/hooks/cdktf-project.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useCdktfProject = void 0;
const ink_1 = require("ink");
const react_1 = require("react");
const lib_1 = require("../../../../lib/");
function useCdktfProject(opts, projectCallback) {
    const { exit } = ink_1.useApp();
    const [id, setID] = react_1.useState(0);
    const [logEntries, setLogEntries] = react_1.useState([]);
    const [returnValue, setReturnValue] = react_1.useState();
    const [status, setStatus] = react_1.useState({ type: "starting" });
    const updateRunningStatus = (project) => {
        const inProgress = project.stacksToRun.filter((s) => s.isRunning);
        const finished = project.stacksToRun.filter((s) => s.isDone);
        const pending = project.stacksToRun.filter((s) => s.isPending);
        setStatus({ type: "running", inProgress, finished, pending });
    };
    react_1.useEffect(() => {
        const project = new lib_1.CdktfProject({
            outDir: opts.outDir,
            synthCommand: opts.synthCommand,
            onUpdate: (update) => {
                if (["synthesizing"].includes(update.type)) {
                    setStatus({ type: "synthesizing" });
                }
                else if (update.type === "waiting for approval") {
                    setStatus({
                        type: "waiting for approval of stack",
                        stackName: update.stackName,
                        approve: update.approve,
                        dismiss: update.dismiss,
                        stop: update.stop,
                    });
                }
                else {
                    updateRunningStatus(project);
                }
            },
            onLog: ({ stackName, message, messageWithConstructPath, }) => {
                setLogEntries((prev) => [
                    ...prev,
                    {
                        id: `${stackName}-${id}`,
                        stackName: stackName,
                        content: messageWithConstructPath
                            ? messageWithConstructPath
                            : message,
                    },
                ]);
                setID((current) => current + 1);
                // We only send logs when the project is running.
                updateRunningStatus(project);
            },
        });
        const onAbort = () => {
            project.hardAbort();
        };
        process.on("SIGINT", onAbort);
        process.on("SIGTERM", onAbort);
        process.on("SIGQUIT", onAbort);
        projectCallback(project)
            .then((value) => {
            setReturnValue(value);
            setStatus({ type: "done" });
        })
            .catch((err) => {
            exit(err);
        });
    }, []);
    return {
        logEntries,
        returnValue,
        status,
    };
}
exports.useCdktfProject = useCdktfProject;
