// @cdktf/provider-helm/lib/data-helm-template.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataHelmTemplate = exports.DataHelmTemplateSetStringList = exports.DataHelmTemplateSetStringOutputReference = exports.dataHelmTemplateSetStringToTerraform = exports.DataHelmTemplateSetSensitiveList = exports.DataHelmTemplateSetSensitiveOutputReference = exports.dataHelmTemplateSetSensitiveToTerraform = exports.DataHelmTemplateSetList = exports.DataHelmTemplateSetOutputReference = exports.dataHelmTemplateSetToTerraform = exports.DataHelmTemplatePostrenderOutputReference = exports.dataHelmTemplatePostrenderToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataHelmTemplatePostrenderToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        binary_path: cdktf.stringToTerraform(struct.binaryPath),
    };
}
exports.dataHelmTemplatePostrenderToTerraform = dataHelmTemplatePostrenderToTerraform;
class DataHelmTemplatePostrenderOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._binaryPath !== undefined) {
            hasAnyValues = true;
            internalValueResult.binaryPath = this._binaryPath;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._binaryPath = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._binaryPath = value.binaryPath;
        }
    }
    get binaryPath() {
        return this.getStringAttribute('binary_path');
    }
    set binaryPath(value) {
        this._binaryPath = value;
    }
    // Temporarily expose input value. Use with caution.
    get binaryPathInput() {
        return this._binaryPath;
    }
}
exports.DataHelmTemplatePostrenderOutputReference = DataHelmTemplatePostrenderOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataHelmTemplatePostrenderOutputReference[_a] = { fqn: "@cdktf/provider-helm.DataHelmTemplatePostrenderOutputReference", version: "2.0.20" };
function dataHelmTemplateSetToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        type: cdktf.stringToTerraform(struct.type),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.dataHelmTemplateSetToTerraform = dataHelmTemplateSetToTerraform;
class DataHelmTemplateSetOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._name = undefined;
            this._type = undefined;
            this._value = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._name = value.name;
            this._type = value.type;
            this._value = value.value;
        }
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    resetType() {
        this._type = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
    get value() {
        return this.getStringAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
}
exports.DataHelmTemplateSetOutputReference = DataHelmTemplateSetOutputReference;
_b = JSII_RTTI_SYMBOL_1;
DataHelmTemplateSetOutputReference[_b] = { fqn: "@cdktf/provider-helm.DataHelmTemplateSetOutputReference", version: "2.0.20" };
class DataHelmTemplateSetList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataHelmTemplateSetOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataHelmTemplateSetList = DataHelmTemplateSetList;
_c = JSII_RTTI_SYMBOL_1;
DataHelmTemplateSetList[_c] = { fqn: "@cdktf/provider-helm.DataHelmTemplateSetList", version: "2.0.20" };
function dataHelmTemplateSetSensitiveToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        type: cdktf.stringToTerraform(struct.type),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.dataHelmTemplateSetSensitiveToTerraform = dataHelmTemplateSetSensitiveToTerraform;
class DataHelmTemplateSetSensitiveOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._name = undefined;
            this._type = undefined;
            this._value = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._name = value.name;
            this._type = value.type;
            this._value = value.value;
        }
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    resetType() {
        this._type = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
    get value() {
        return this.getStringAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
}
exports.DataHelmTemplateSetSensitiveOutputReference = DataHelmTemplateSetSensitiveOutputReference;
_d = JSII_RTTI_SYMBOL_1;
DataHelmTemplateSetSensitiveOutputReference[_d] = { fqn: "@cdktf/provider-helm.DataHelmTemplateSetSensitiveOutputReference", version: "2.0.20" };
class DataHelmTemplateSetSensitiveList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataHelmTemplateSetSensitiveOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataHelmTemplateSetSensitiveList = DataHelmTemplateSetSensitiveList;
_e = JSII_RTTI_SYMBOL_1;
DataHelmTemplateSetSensitiveList[_e] = { fqn: "@cdktf/provider-helm.DataHelmTemplateSetSensitiveList", version: "2.0.20" };
function dataHelmTemplateSetStringToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.dataHelmTemplateSetStringToTerraform = dataHelmTemplateSetStringToTerraform;
class DataHelmTemplateSetStringOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._name = undefined;
            this._value = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._name = value.name;
            this._value = value.value;
        }
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get value() {
        return this.getStringAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
}
exports.DataHelmTemplateSetStringOutputReference = DataHelmTemplateSetStringOutputReference;
_f = JSII_RTTI_SYMBOL_1;
DataHelmTemplateSetStringOutputReference[_f] = { fqn: "@cdktf/provider-helm.DataHelmTemplateSetStringOutputReference", version: "2.0.20" };
class DataHelmTemplateSetStringList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataHelmTemplateSetStringOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataHelmTemplateSetStringList = DataHelmTemplateSetStringList;
_g = JSII_RTTI_SYMBOL_1;
DataHelmTemplateSetStringList[_g] = { fqn: "@cdktf/provider-helm.DataHelmTemplateSetStringList", version: "2.0.20" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/helm/d/template helm_template}
*/
class DataHelmTemplate extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/helm/d/template helm_template} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataHelmTemplateConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'helm_template',
            terraformGeneratorMetadata: {
                providerName: 'helm',
                providerVersion: '2.6.0',
                providerVersionConstraint: '~> 2.3'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // postrender - computed: false, optional: true, required: false
        this._postrender = new DataHelmTemplatePostrenderOutputReference(this, "postrender");
        // set - computed: false, optional: true, required: false
        this._set = new DataHelmTemplateSetList(this, "set", true);
        // set_sensitive - computed: false, optional: true, required: false
        this._setSensitive = new DataHelmTemplateSetSensitiveList(this, "set_sensitive", true);
        // set_string - computed: false, optional: true, required: false
        this._setString = new DataHelmTemplateSetStringList(this, "set_string", true);
        this._apiVersions = config.apiVersions;
        this._atomic = config.atomic;
        this._chart = config.chart;
        this._createNamespace = config.createNamespace;
        this._dependencyUpdate = config.dependencyUpdate;
        this._description = config.description;
        this._devel = config.devel;
        this._disableOpenapiValidation = config.disableOpenapiValidation;
        this._disableWebhooks = config.disableWebhooks;
        this._id = config.id;
        this._includeCrds = config.includeCrds;
        this._isUpgrade = config.isUpgrade;
        this._keyring = config.keyring;
        this._manifest = config.manifest;
        this._manifests = config.manifests;
        this._name = config.name;
        this._namespace = config.namespace;
        this._notes = config.notes;
        this._passCredentials = config.passCredentials;
        this._renderSubchartNotes = config.renderSubchartNotes;
        this._replace = config.replace;
        this._repository = config.repository;
        this._repositoryCaFile = config.repositoryCaFile;
        this._repositoryCertFile = config.repositoryCertFile;
        this._repositoryKeyFile = config.repositoryKeyFile;
        this._repositoryPassword = config.repositoryPassword;
        this._repositoryUsername = config.repositoryUsername;
        this._resetValues = config.resetValues;
        this._reuseValues = config.reuseValues;
        this._showOnly = config.showOnly;
        this._skipCrds = config.skipCrds;
        this._skipTests = config.skipTests;
        this._timeout = config.timeout;
        this._validate = config.validate;
        this._values = config.values;
        this._verify = config.verify;
        this._version = config.version;
        this._wait = config.wait;
        this._postrender.internalValue = config.postrender;
        this._set.internalValue = config.set;
        this._setSensitive.internalValue = config.setSensitive;
        this._setString.internalValue = config.setString;
    }
    get apiVersions() {
        return this.getListAttribute('api_versions');
    }
    set apiVersions(value) {
        this._apiVersions = value;
    }
    resetApiVersions() {
        this._apiVersions = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get apiVersionsInput() {
        return this._apiVersions;
    }
    get atomic() {
        return this.getBooleanAttribute('atomic');
    }
    set atomic(value) {
        this._atomic = value;
    }
    resetAtomic() {
        this._atomic = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get atomicInput() {
        return this._atomic;
    }
    get chart() {
        return this.getStringAttribute('chart');
    }
    set chart(value) {
        this._chart = value;
    }
    // Temporarily expose input value. Use with caution.
    get chartInput() {
        return this._chart;
    }
    get createNamespace() {
        return this.getBooleanAttribute('create_namespace');
    }
    set createNamespace(value) {
        this._createNamespace = value;
    }
    resetCreateNamespace() {
        this._createNamespace = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createNamespaceInput() {
        return this._createNamespace;
    }
    get dependencyUpdate() {
        return this.getBooleanAttribute('dependency_update');
    }
    set dependencyUpdate(value) {
        this._dependencyUpdate = value;
    }
    resetDependencyUpdate() {
        this._dependencyUpdate = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dependencyUpdateInput() {
        return this._dependencyUpdate;
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get devel() {
        return this.getBooleanAttribute('devel');
    }
    set devel(value) {
        this._devel = value;
    }
    resetDevel() {
        this._devel = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get develInput() {
        return this._devel;
    }
    get disableOpenapiValidation() {
        return this.getBooleanAttribute('disable_openapi_validation');
    }
    set disableOpenapiValidation(value) {
        this._disableOpenapiValidation = value;
    }
    resetDisableOpenapiValidation() {
        this._disableOpenapiValidation = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get disableOpenapiValidationInput() {
        return this._disableOpenapiValidation;
    }
    get disableWebhooks() {
        return this.getBooleanAttribute('disable_webhooks');
    }
    set disableWebhooks(value) {
        this._disableWebhooks = value;
    }
    resetDisableWebhooks() {
        this._disableWebhooks = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get disableWebhooksInput() {
        return this._disableWebhooks;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get includeCrds() {
        return this.getBooleanAttribute('include_crds');
    }
    set includeCrds(value) {
        this._includeCrds = value;
    }
    resetIncludeCrds() {
        this._includeCrds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get includeCrdsInput() {
        return this._includeCrds;
    }
    get isUpgrade() {
        return this.getBooleanAttribute('is_upgrade');
    }
    set isUpgrade(value) {
        this._isUpgrade = value;
    }
    resetIsUpgrade() {
        this._isUpgrade = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get isUpgradeInput() {
        return this._isUpgrade;
    }
    get keyring() {
        return this.getStringAttribute('keyring');
    }
    set keyring(value) {
        this._keyring = value;
    }
    resetKeyring() {
        this._keyring = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get keyringInput() {
        return this._keyring;
    }
    get manifest() {
        return this.getStringAttribute('manifest');
    }
    set manifest(value) {
        this._manifest = value;
    }
    resetManifest() {
        this._manifest = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get manifestInput() {
        return this._manifest;
    }
    get manifests() {
        return this.getStringMapAttribute('manifests');
    }
    set manifests(value) {
        this._manifests = value;
    }
    resetManifests() {
        this._manifests = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get manifestsInput() {
        return this._manifests;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get namespace() {
        return this.getStringAttribute('namespace');
    }
    set namespace(value) {
        this._namespace = value;
    }
    resetNamespace() {
        this._namespace = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get namespaceInput() {
        return this._namespace;
    }
    get notes() {
        return this.getStringAttribute('notes');
    }
    set notes(value) {
        this._notes = value;
    }
    resetNotes() {
        this._notes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get notesInput() {
        return this._notes;
    }
    get passCredentials() {
        return this.getBooleanAttribute('pass_credentials');
    }
    set passCredentials(value) {
        this._passCredentials = value;
    }
    resetPassCredentials() {
        this._passCredentials = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get passCredentialsInput() {
        return this._passCredentials;
    }
    get renderSubchartNotes() {
        return this.getBooleanAttribute('render_subchart_notes');
    }
    set renderSubchartNotes(value) {
        this._renderSubchartNotes = value;
    }
    resetRenderSubchartNotes() {
        this._renderSubchartNotes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get renderSubchartNotesInput() {
        return this._renderSubchartNotes;
    }
    get replace() {
        return this.getBooleanAttribute('replace');
    }
    set replace(value) {
        this._replace = value;
    }
    resetReplace() {
        this._replace = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get replaceInput() {
        return this._replace;
    }
    get repository() {
        return this.getStringAttribute('repository');
    }
    set repository(value) {
        this._repository = value;
    }
    resetRepository() {
        this._repository = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get repositoryInput() {
        return this._repository;
    }
    get repositoryCaFile() {
        return this.getStringAttribute('repository_ca_file');
    }
    set repositoryCaFile(value) {
        this._repositoryCaFile = value;
    }
    resetRepositoryCaFile() {
        this._repositoryCaFile = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get repositoryCaFileInput() {
        return this._repositoryCaFile;
    }
    get repositoryCertFile() {
        return this.getStringAttribute('repository_cert_file');
    }
    set repositoryCertFile(value) {
        this._repositoryCertFile = value;
    }
    resetRepositoryCertFile() {
        this._repositoryCertFile = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get repositoryCertFileInput() {
        return this._repositoryCertFile;
    }
    get repositoryKeyFile() {
        return this.getStringAttribute('repository_key_file');
    }
    set repositoryKeyFile(value) {
        this._repositoryKeyFile = value;
    }
    resetRepositoryKeyFile() {
        this._repositoryKeyFile = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get repositoryKeyFileInput() {
        return this._repositoryKeyFile;
    }
    get repositoryPassword() {
        return this.getStringAttribute('repository_password');
    }
    set repositoryPassword(value) {
        this._repositoryPassword = value;
    }
    resetRepositoryPassword() {
        this._repositoryPassword = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get repositoryPasswordInput() {
        return this._repositoryPassword;
    }
    get repositoryUsername() {
        return this.getStringAttribute('repository_username');
    }
    set repositoryUsername(value) {
        this._repositoryUsername = value;
    }
    resetRepositoryUsername() {
        this._repositoryUsername = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get repositoryUsernameInput() {
        return this._repositoryUsername;
    }
    get resetValues() {
        return this.getBooleanAttribute('reset_values');
    }
    set resetValues(value) {
        this._resetValues = value;
    }
    resetResetValues() {
        this._resetValues = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get resetValuesInput() {
        return this._resetValues;
    }
    get reuseValues() {
        return this.getBooleanAttribute('reuse_values');
    }
    set reuseValues(value) {
        this._reuseValues = value;
    }
    resetReuseValues() {
        this._reuseValues = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get reuseValuesInput() {
        return this._reuseValues;
    }
    get showOnly() {
        return this.getListAttribute('show_only');
    }
    set showOnly(value) {
        this._showOnly = value;
    }
    resetShowOnly() {
        this._showOnly = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get showOnlyInput() {
        return this._showOnly;
    }
    get skipCrds() {
        return this.getBooleanAttribute('skip_crds');
    }
    set skipCrds(value) {
        this._skipCrds = value;
    }
    resetSkipCrds() {
        this._skipCrds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get skipCrdsInput() {
        return this._skipCrds;
    }
    get skipTests() {
        return this.getBooleanAttribute('skip_tests');
    }
    set skipTests(value) {
        this._skipTests = value;
    }
    resetSkipTests() {
        this._skipTests = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get skipTestsInput() {
        return this._skipTests;
    }
    get timeout() {
        return this.getNumberAttribute('timeout');
    }
    set timeout(value) {
        this._timeout = value;
    }
    resetTimeout() {
        this._timeout = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutInput() {
        return this._timeout;
    }
    get validate() {
        return this.getBooleanAttribute('validate');
    }
    set validate(value) {
        this._validate = value;
    }
    resetValidate() {
        this._validate = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get validateInput() {
        return this._validate;
    }
    get values() {
        return this.getListAttribute('values');
    }
    set values(value) {
        this._values = value;
    }
    resetTfValues() {
        this._values = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get valuesInput() {
        return this._values;
    }
    get verify() {
        return this.getBooleanAttribute('verify');
    }
    set verify(value) {
        this._verify = value;
    }
    resetVerify() {
        this._verify = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get verifyInput() {
        return this._verify;
    }
    get version() {
        return this.getStringAttribute('version');
    }
    set version(value) {
        this._version = value;
    }
    resetVersion() {
        this._version = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get versionInput() {
        return this._version;
    }
    get wait() {
        return this.getBooleanAttribute('wait');
    }
    set wait(value) {
        this._wait = value;
    }
    resetWait() {
        this._wait = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get waitInput() {
        return this._wait;
    }
    get postrender() {
        return this._postrender;
    }
    putPostrender(value) {
        this._postrender.internalValue = value;
    }
    resetPostrender() {
        this._postrender.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get postrenderInput() {
        return this._postrender.internalValue;
    }
    get set() {
        return this._set;
    }
    putSet(value) {
        this._set.internalValue = value;
    }
    resetSet() {
        this._set.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get setInput() {
        return this._set.internalValue;
    }
    get setSensitive() {
        return this._setSensitive;
    }
    putSetSensitive(value) {
        this._setSensitive.internalValue = value;
    }
    resetSetSensitive() {
        this._setSensitive.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get setSensitiveInput() {
        return this._setSensitive.internalValue;
    }
    get setString() {
        return this._setString;
    }
    putSetString(value) {
        this._setString.internalValue = value;
    }
    resetSetString() {
        this._setString.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get setStringInput() {
        return this._setString.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            api_versions: cdktf.listMapper(cdktf.stringToTerraform, false)(this._apiVersions),
            atomic: cdktf.booleanToTerraform(this._atomic),
            chart: cdktf.stringToTerraform(this._chart),
            create_namespace: cdktf.booleanToTerraform(this._createNamespace),
            dependency_update: cdktf.booleanToTerraform(this._dependencyUpdate),
            description: cdktf.stringToTerraform(this._description),
            devel: cdktf.booleanToTerraform(this._devel),
            disable_openapi_validation: cdktf.booleanToTerraform(this._disableOpenapiValidation),
            disable_webhooks: cdktf.booleanToTerraform(this._disableWebhooks),
            id: cdktf.stringToTerraform(this._id),
            include_crds: cdktf.booleanToTerraform(this._includeCrds),
            is_upgrade: cdktf.booleanToTerraform(this._isUpgrade),
            keyring: cdktf.stringToTerraform(this._keyring),
            manifest: cdktf.stringToTerraform(this._manifest),
            manifests: cdktf.hashMapper(cdktf.stringToTerraform)(this._manifests),
            name: cdktf.stringToTerraform(this._name),
            namespace: cdktf.stringToTerraform(this._namespace),
            notes: cdktf.stringToTerraform(this._notes),
            pass_credentials: cdktf.booleanToTerraform(this._passCredentials),
            render_subchart_notes: cdktf.booleanToTerraform(this._renderSubchartNotes),
            replace: cdktf.booleanToTerraform(this._replace),
            repository: cdktf.stringToTerraform(this._repository),
            repository_ca_file: cdktf.stringToTerraform(this._repositoryCaFile),
            repository_cert_file: cdktf.stringToTerraform(this._repositoryCertFile),
            repository_key_file: cdktf.stringToTerraform(this._repositoryKeyFile),
            repository_password: cdktf.stringToTerraform(this._repositoryPassword),
            repository_username: cdktf.stringToTerraform(this._repositoryUsername),
            reset_values: cdktf.booleanToTerraform(this._resetValues),
            reuse_values: cdktf.booleanToTerraform(this._reuseValues),
            show_only: cdktf.listMapper(cdktf.stringToTerraform, false)(this._showOnly),
            skip_crds: cdktf.booleanToTerraform(this._skipCrds),
            skip_tests: cdktf.booleanToTerraform(this._skipTests),
            timeout: cdktf.numberToTerraform(this._timeout),
            validate: cdktf.booleanToTerraform(this._validate),
            values: cdktf.listMapper(cdktf.stringToTerraform, false)(this._values),
            verify: cdktf.booleanToTerraform(this._verify),
            version: cdktf.stringToTerraform(this._version),
            wait: cdktf.booleanToTerraform(this._wait),
            postrender: dataHelmTemplatePostrenderToTerraform(this._postrender.internalValue),
            set: cdktf.listMapper(dataHelmTemplateSetToTerraform, true)(this._set.internalValue),
            set_sensitive: cdktf.listMapper(dataHelmTemplateSetSensitiveToTerraform, true)(this._setSensitive.internalValue),
            set_string: cdktf.listMapper(dataHelmTemplateSetStringToTerraform, true)(this._setString.internalValue),
        };
    }
}
exports.DataHelmTemplate = DataHelmTemplate;
_h = JSII_RTTI_SYMBOL_1;
DataHelmTemplate[_h] = { fqn: "@cdktf/provider-helm.DataHelmTemplate", version: "2.0.20" };
// =================
// STATIC PROPERTIES
// =================
DataHelmTemplate.tfResourceType = "helm_template";
