// @cdktf/provider-helm/lib/release.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Release = exports.ReleaseSetSensitiveList = exports.ReleaseSetSensitiveOutputReference = exports.releaseSetSensitiveToTerraform = exports.ReleaseSetList = exports.ReleaseSetOutputReference = exports.releaseSetToTerraform = exports.ReleasePostrenderOutputReference = exports.releasePostrenderToTerraform = exports.ReleaseMetadataList = exports.ReleaseMetadataOutputReference = exports.releaseMetadataToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function releaseMetadataToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.releaseMetadataToTerraform = releaseMetadataToTerraform;
class ReleaseMetadataOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // app_version - computed: true, optional: false, required: false
    get appVersion() {
        return this.getStringAttribute('app_version');
    }
    // chart - computed: true, optional: false, required: false
    get chart() {
        return this.getStringAttribute('chart');
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // namespace - computed: true, optional: false, required: false
    get namespace() {
        return this.getStringAttribute('namespace');
    }
    // revision - computed: true, optional: false, required: false
    get revision() {
        return this.getNumberAttribute('revision');
    }
    // values - computed: true, optional: false, required: false
    get values() {
        return this.getStringAttribute('values');
    }
    // version - computed: true, optional: false, required: false
    get version() {
        return this.getStringAttribute('version');
    }
}
exports.ReleaseMetadataOutputReference = ReleaseMetadataOutputReference;
_a = JSII_RTTI_SYMBOL_1;
ReleaseMetadataOutputReference[_a] = { fqn: "@cdktf/provider-helm.ReleaseMetadataOutputReference", version: "2.0.20" };
class ReleaseMetadataList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ReleaseMetadataOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ReleaseMetadataList = ReleaseMetadataList;
_b = JSII_RTTI_SYMBOL_1;
ReleaseMetadataList[_b] = { fqn: "@cdktf/provider-helm.ReleaseMetadataList", version: "2.0.20" };
function releasePostrenderToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        args: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.args),
        binary_path: cdktf.stringToTerraform(struct.binaryPath),
    };
}
exports.releasePostrenderToTerraform = releasePostrenderToTerraform;
class ReleasePostrenderOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._args !== undefined) {
            hasAnyValues = true;
            internalValueResult.args = this._args;
        }
        if (this._binaryPath !== undefined) {
            hasAnyValues = true;
            internalValueResult.binaryPath = this._binaryPath;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._args = undefined;
            this._binaryPath = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._args = value.args;
            this._binaryPath = value.binaryPath;
        }
    }
    get args() {
        return this.getListAttribute('args');
    }
    set args(value) {
        this._args = value;
    }
    resetArgs() {
        this._args = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get argsInput() {
        return this._args;
    }
    get binaryPath() {
        return this.getStringAttribute('binary_path');
    }
    set binaryPath(value) {
        this._binaryPath = value;
    }
    // Temporarily expose input value. Use with caution.
    get binaryPathInput() {
        return this._binaryPath;
    }
}
exports.ReleasePostrenderOutputReference = ReleasePostrenderOutputReference;
_c = JSII_RTTI_SYMBOL_1;
ReleasePostrenderOutputReference[_c] = { fqn: "@cdktf/provider-helm.ReleasePostrenderOutputReference", version: "2.0.20" };
function releaseSetToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        type: cdktf.stringToTerraform(struct.type),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.releaseSetToTerraform = releaseSetToTerraform;
class ReleaseSetOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._name = undefined;
            this._type = undefined;
            this._value = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._name = value.name;
            this._type = value.type;
            this._value = value.value;
        }
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    resetType() {
        this._type = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
    get value() {
        return this.getStringAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
}
exports.ReleaseSetOutputReference = ReleaseSetOutputReference;
_d = JSII_RTTI_SYMBOL_1;
ReleaseSetOutputReference[_d] = { fqn: "@cdktf/provider-helm.ReleaseSetOutputReference", version: "2.0.20" };
class ReleaseSetList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ReleaseSetOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ReleaseSetList = ReleaseSetList;
_e = JSII_RTTI_SYMBOL_1;
ReleaseSetList[_e] = { fqn: "@cdktf/provider-helm.ReleaseSetList", version: "2.0.20" };
function releaseSetSensitiveToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        type: cdktf.stringToTerraform(struct.type),
        value: cdktf.stringToTerraform(struct.value),
    };
}
exports.releaseSetSensitiveToTerraform = releaseSetSensitiveToTerraform;
class ReleaseSetSensitiveOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        if (this._value !== undefined) {
            hasAnyValues = true;
            internalValueResult.value = this._value;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._name = undefined;
            this._type = undefined;
            this._value = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._name = value.name;
            this._type = value.type;
            this._value = value.value;
        }
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    resetType() {
        this._type = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
    get value() {
        return this.getStringAttribute('value');
    }
    set value(value) {
        this._value = value;
    }
    // Temporarily expose input value. Use with caution.
    get valueInput() {
        return this._value;
    }
}
exports.ReleaseSetSensitiveOutputReference = ReleaseSetSensitiveOutputReference;
_f = JSII_RTTI_SYMBOL_1;
ReleaseSetSensitiveOutputReference[_f] = { fqn: "@cdktf/provider-helm.ReleaseSetSensitiveOutputReference", version: "2.0.20" };
class ReleaseSetSensitiveList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ReleaseSetSensitiveOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ReleaseSetSensitiveList = ReleaseSetSensitiveList;
_g = JSII_RTTI_SYMBOL_1;
ReleaseSetSensitiveList[_g] = { fqn: "@cdktf/provider-helm.ReleaseSetSensitiveList", version: "2.0.20" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/helm/r/release helm_release}
*/
class Release extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/helm/r/release helm_release} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options ReleaseConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'helm_release',
            terraformGeneratorMetadata: {
                providerName: 'helm',
                providerVersion: '2.6.0',
                providerVersionConstraint: '~> 2.3'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // metadata - computed: true, optional: false, required: false
        this._metadata = new ReleaseMetadataList(this, "metadata", false);
        // postrender - computed: false, optional: true, required: false
        this._postrender = new ReleasePostrenderOutputReference(this, "postrender");
        // set - computed: false, optional: true, required: false
        this._set = new ReleaseSetList(this, "set", true);
        // set_sensitive - computed: false, optional: true, required: false
        this._setSensitive = new ReleaseSetSensitiveList(this, "set_sensitive", true);
        this._atomic = config.atomic;
        this._chart = config.chart;
        this._cleanupOnFail = config.cleanupOnFail;
        this._createNamespace = config.createNamespace;
        this._dependencyUpdate = config.dependencyUpdate;
        this._description = config.description;
        this._devel = config.devel;
        this._disableCrdHooks = config.disableCrdHooks;
        this._disableOpenapiValidation = config.disableOpenapiValidation;
        this._disableWebhooks = config.disableWebhooks;
        this._forceUpdate = config.forceUpdate;
        this._id = config.id;
        this._keyring = config.keyring;
        this._lint = config.lint;
        this._maxHistory = config.maxHistory;
        this._name = config.name;
        this._namespace = config.namespace;
        this._passCredentials = config.passCredentials;
        this._recreatePods = config.recreatePods;
        this._renderSubchartNotes = config.renderSubchartNotes;
        this._replace = config.replace;
        this._repository = config.repository;
        this._repositoryCaFile = config.repositoryCaFile;
        this._repositoryCertFile = config.repositoryCertFile;
        this._repositoryKeyFile = config.repositoryKeyFile;
        this._repositoryPassword = config.repositoryPassword;
        this._repositoryUsername = config.repositoryUsername;
        this._resetValues = config.resetValues;
        this._reuseValues = config.reuseValues;
        this._skipCrds = config.skipCrds;
        this._timeout = config.timeout;
        this._values = config.values;
        this._verify = config.verify;
        this._version = config.version;
        this._wait = config.wait;
        this._waitForJobs = config.waitForJobs;
        this._postrender.internalValue = config.postrender;
        this._set.internalValue = config.set;
        this._setSensitive.internalValue = config.setSensitive;
    }
    get atomic() {
        return this.getBooleanAttribute('atomic');
    }
    set atomic(value) {
        this._atomic = value;
    }
    resetAtomic() {
        this._atomic = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get atomicInput() {
        return this._atomic;
    }
    get chart() {
        return this.getStringAttribute('chart');
    }
    set chart(value) {
        this._chart = value;
    }
    // Temporarily expose input value. Use with caution.
    get chartInput() {
        return this._chart;
    }
    get cleanupOnFail() {
        return this.getBooleanAttribute('cleanup_on_fail');
    }
    set cleanupOnFail(value) {
        this._cleanupOnFail = value;
    }
    resetCleanupOnFail() {
        this._cleanupOnFail = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get cleanupOnFailInput() {
        return this._cleanupOnFail;
    }
    get createNamespace() {
        return this.getBooleanAttribute('create_namespace');
    }
    set createNamespace(value) {
        this._createNamespace = value;
    }
    resetCreateNamespace() {
        this._createNamespace = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createNamespaceInput() {
        return this._createNamespace;
    }
    get dependencyUpdate() {
        return this.getBooleanAttribute('dependency_update');
    }
    set dependencyUpdate(value) {
        this._dependencyUpdate = value;
    }
    resetDependencyUpdate() {
        this._dependencyUpdate = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dependencyUpdateInput() {
        return this._dependencyUpdate;
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get devel() {
        return this.getBooleanAttribute('devel');
    }
    set devel(value) {
        this._devel = value;
    }
    resetDevel() {
        this._devel = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get develInput() {
        return this._devel;
    }
    get disableCrdHooks() {
        return this.getBooleanAttribute('disable_crd_hooks');
    }
    set disableCrdHooks(value) {
        this._disableCrdHooks = value;
    }
    resetDisableCrdHooks() {
        this._disableCrdHooks = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get disableCrdHooksInput() {
        return this._disableCrdHooks;
    }
    get disableOpenapiValidation() {
        return this.getBooleanAttribute('disable_openapi_validation');
    }
    set disableOpenapiValidation(value) {
        this._disableOpenapiValidation = value;
    }
    resetDisableOpenapiValidation() {
        this._disableOpenapiValidation = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get disableOpenapiValidationInput() {
        return this._disableOpenapiValidation;
    }
    get disableWebhooks() {
        return this.getBooleanAttribute('disable_webhooks');
    }
    set disableWebhooks(value) {
        this._disableWebhooks = value;
    }
    resetDisableWebhooks() {
        this._disableWebhooks = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get disableWebhooksInput() {
        return this._disableWebhooks;
    }
    get forceUpdate() {
        return this.getBooleanAttribute('force_update');
    }
    set forceUpdate(value) {
        this._forceUpdate = value;
    }
    resetForceUpdate() {
        this._forceUpdate = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get forceUpdateInput() {
        return this._forceUpdate;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get keyring() {
        return this.getStringAttribute('keyring');
    }
    set keyring(value) {
        this._keyring = value;
    }
    resetKeyring() {
        this._keyring = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get keyringInput() {
        return this._keyring;
    }
    get lint() {
        return this.getBooleanAttribute('lint');
    }
    set lint(value) {
        this._lint = value;
    }
    resetLint() {
        this._lint = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get lintInput() {
        return this._lint;
    }
    // manifest - computed: true, optional: false, required: false
    get manifest() {
        return this.getStringAttribute('manifest');
    }
    get maxHistory() {
        return this.getNumberAttribute('max_history');
    }
    set maxHistory(value) {
        this._maxHistory = value;
    }
    resetMaxHistory() {
        this._maxHistory = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maxHistoryInput() {
        return this._maxHistory;
    }
    get metadata() {
        return this._metadata;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get namespace() {
        return this.getStringAttribute('namespace');
    }
    set namespace(value) {
        this._namespace = value;
    }
    resetNamespace() {
        this._namespace = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get namespaceInput() {
        return this._namespace;
    }
    get passCredentials() {
        return this.getBooleanAttribute('pass_credentials');
    }
    set passCredentials(value) {
        this._passCredentials = value;
    }
    resetPassCredentials() {
        this._passCredentials = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get passCredentialsInput() {
        return this._passCredentials;
    }
    get recreatePods() {
        return this.getBooleanAttribute('recreate_pods');
    }
    set recreatePods(value) {
        this._recreatePods = value;
    }
    resetRecreatePods() {
        this._recreatePods = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get recreatePodsInput() {
        return this._recreatePods;
    }
    get renderSubchartNotes() {
        return this.getBooleanAttribute('render_subchart_notes');
    }
    set renderSubchartNotes(value) {
        this._renderSubchartNotes = value;
    }
    resetRenderSubchartNotes() {
        this._renderSubchartNotes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get renderSubchartNotesInput() {
        return this._renderSubchartNotes;
    }
    get replace() {
        return this.getBooleanAttribute('replace');
    }
    set replace(value) {
        this._replace = value;
    }
    resetReplace() {
        this._replace = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get replaceInput() {
        return this._replace;
    }
    get repository() {
        return this.getStringAttribute('repository');
    }
    set repository(value) {
        this._repository = value;
    }
    resetRepository() {
        this._repository = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get repositoryInput() {
        return this._repository;
    }
    get repositoryCaFile() {
        return this.getStringAttribute('repository_ca_file');
    }
    set repositoryCaFile(value) {
        this._repositoryCaFile = value;
    }
    resetRepositoryCaFile() {
        this._repositoryCaFile = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get repositoryCaFileInput() {
        return this._repositoryCaFile;
    }
    get repositoryCertFile() {
        return this.getStringAttribute('repository_cert_file');
    }
    set repositoryCertFile(value) {
        this._repositoryCertFile = value;
    }
    resetRepositoryCertFile() {
        this._repositoryCertFile = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get repositoryCertFileInput() {
        return this._repositoryCertFile;
    }
    get repositoryKeyFile() {
        return this.getStringAttribute('repository_key_file');
    }
    set repositoryKeyFile(value) {
        this._repositoryKeyFile = value;
    }
    resetRepositoryKeyFile() {
        this._repositoryKeyFile = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get repositoryKeyFileInput() {
        return this._repositoryKeyFile;
    }
    get repositoryPassword() {
        return this.getStringAttribute('repository_password');
    }
    set repositoryPassword(value) {
        this._repositoryPassword = value;
    }
    resetRepositoryPassword() {
        this._repositoryPassword = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get repositoryPasswordInput() {
        return this._repositoryPassword;
    }
    get repositoryUsername() {
        return this.getStringAttribute('repository_username');
    }
    set repositoryUsername(value) {
        this._repositoryUsername = value;
    }
    resetRepositoryUsername() {
        this._repositoryUsername = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get repositoryUsernameInput() {
        return this._repositoryUsername;
    }
    get resetValues() {
        return this.getBooleanAttribute('reset_values');
    }
    set resetValues(value) {
        this._resetValues = value;
    }
    resetResetValues() {
        this._resetValues = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get resetValuesInput() {
        return this._resetValues;
    }
    get reuseValues() {
        return this.getBooleanAttribute('reuse_values');
    }
    set reuseValues(value) {
        this._reuseValues = value;
    }
    resetReuseValues() {
        this._reuseValues = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get reuseValuesInput() {
        return this._reuseValues;
    }
    get skipCrds() {
        return this.getBooleanAttribute('skip_crds');
    }
    set skipCrds(value) {
        this._skipCrds = value;
    }
    resetSkipCrds() {
        this._skipCrds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get skipCrdsInput() {
        return this._skipCrds;
    }
    // status - computed: true, optional: false, required: false
    get status() {
        return this.getStringAttribute('status');
    }
    get timeout() {
        return this.getNumberAttribute('timeout');
    }
    set timeout(value) {
        this._timeout = value;
    }
    resetTimeout() {
        this._timeout = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutInput() {
        return this._timeout;
    }
    get values() {
        return this.getListAttribute('values');
    }
    set values(value) {
        this._values = value;
    }
    resetTfValues() {
        this._values = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get valuesInput() {
        return this._values;
    }
    get verify() {
        return this.getBooleanAttribute('verify');
    }
    set verify(value) {
        this._verify = value;
    }
    resetVerify() {
        this._verify = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get verifyInput() {
        return this._verify;
    }
    get version() {
        return this.getStringAttribute('version');
    }
    set version(value) {
        this._version = value;
    }
    resetVersion() {
        this._version = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get versionInput() {
        return this._version;
    }
    get wait() {
        return this.getBooleanAttribute('wait');
    }
    set wait(value) {
        this._wait = value;
    }
    resetWait() {
        this._wait = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get waitInput() {
        return this._wait;
    }
    get waitForJobs() {
        return this.getBooleanAttribute('wait_for_jobs');
    }
    set waitForJobs(value) {
        this._waitForJobs = value;
    }
    resetWaitForJobs() {
        this._waitForJobs = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get waitForJobsInput() {
        return this._waitForJobs;
    }
    get postrender() {
        return this._postrender;
    }
    putPostrender(value) {
        this._postrender.internalValue = value;
    }
    resetPostrender() {
        this._postrender.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get postrenderInput() {
        return this._postrender.internalValue;
    }
    get set() {
        return this._set;
    }
    putSet(value) {
        this._set.internalValue = value;
    }
    resetSet() {
        this._set.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get setInput() {
        return this._set.internalValue;
    }
    get setSensitive() {
        return this._setSensitive;
    }
    putSetSensitive(value) {
        this._setSensitive.internalValue = value;
    }
    resetSetSensitive() {
        this._setSensitive.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get setSensitiveInput() {
        return this._setSensitive.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            atomic: cdktf.booleanToTerraform(this._atomic),
            chart: cdktf.stringToTerraform(this._chart),
            cleanup_on_fail: cdktf.booleanToTerraform(this._cleanupOnFail),
            create_namespace: cdktf.booleanToTerraform(this._createNamespace),
            dependency_update: cdktf.booleanToTerraform(this._dependencyUpdate),
            description: cdktf.stringToTerraform(this._description),
            devel: cdktf.booleanToTerraform(this._devel),
            disable_crd_hooks: cdktf.booleanToTerraform(this._disableCrdHooks),
            disable_openapi_validation: cdktf.booleanToTerraform(this._disableOpenapiValidation),
            disable_webhooks: cdktf.booleanToTerraform(this._disableWebhooks),
            force_update: cdktf.booleanToTerraform(this._forceUpdate),
            id: cdktf.stringToTerraform(this._id),
            keyring: cdktf.stringToTerraform(this._keyring),
            lint: cdktf.booleanToTerraform(this._lint),
            max_history: cdktf.numberToTerraform(this._maxHistory),
            name: cdktf.stringToTerraform(this._name),
            namespace: cdktf.stringToTerraform(this._namespace),
            pass_credentials: cdktf.booleanToTerraform(this._passCredentials),
            recreate_pods: cdktf.booleanToTerraform(this._recreatePods),
            render_subchart_notes: cdktf.booleanToTerraform(this._renderSubchartNotes),
            replace: cdktf.booleanToTerraform(this._replace),
            repository: cdktf.stringToTerraform(this._repository),
            repository_ca_file: cdktf.stringToTerraform(this._repositoryCaFile),
            repository_cert_file: cdktf.stringToTerraform(this._repositoryCertFile),
            repository_key_file: cdktf.stringToTerraform(this._repositoryKeyFile),
            repository_password: cdktf.stringToTerraform(this._repositoryPassword),
            repository_username: cdktf.stringToTerraform(this._repositoryUsername),
            reset_values: cdktf.booleanToTerraform(this._resetValues),
            reuse_values: cdktf.booleanToTerraform(this._reuseValues),
            skip_crds: cdktf.booleanToTerraform(this._skipCrds),
            timeout: cdktf.numberToTerraform(this._timeout),
            values: cdktf.listMapper(cdktf.stringToTerraform, false)(this._values),
            verify: cdktf.booleanToTerraform(this._verify),
            version: cdktf.stringToTerraform(this._version),
            wait: cdktf.booleanToTerraform(this._wait),
            wait_for_jobs: cdktf.booleanToTerraform(this._waitForJobs),
            postrender: releasePostrenderToTerraform(this._postrender.internalValue),
            set: cdktf.listMapper(releaseSetToTerraform, true)(this._set.internalValue),
            set_sensitive: cdktf.listMapper(releaseSetSensitiveToTerraform, true)(this._setSensitive.internalValue),
        };
    }
}
exports.Release = Release;
_h = JSII_RTTI_SYMBOL_1;
Release[_h] = { fqn: "@cdktf/provider-helm.Release", version: "2.0.20" };
// =================
// STATIC PROPERTIES
// =================
Release.tfResourceType = "helm_release";
